apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/json/MapJsonReader.kt

```diff
@@ -23,285 +24,330 @@ import com.apollographql.apollo3.api.json.internal.toLongExact
  *
  * @param root the root [Map] to read from
  *
- * Because it's mostly used internally the checks/verifications are very slim.
  * To read from a [okio.BufferedSource], see also [BufferedSourceJsonReader]
  */
 class MapJsonReader(val root: Map<String, Any?>) : JsonReader {
-  // Like a Map but easier to access sequentially
-  // Could be optimized
-  private class OrderedMap(val entries: List<Entry>)
-  private class Entry(val key: String, val value: Any?)
 
-  private val dataStack = arrayOfNulls<Any>(MAX_STACK_SIZE)
-  private val indexStack = IntArray(MAX_STACK_SIZE)
-  private val nameIndexStack = IntArray(MAX_STACK_SIZE)
+  private var peekedToken: JsonReader.Token = JsonReader.Token.END_OBJECT
 
-  private var stackSize = 0
+  /**
+   * Depending what [peekedToken] is, [peekedData] can be safely cast to a Map, Entry
+   * or other values
+   */
+  private var peekedData: Any? = null
 
-  // Will be non-null if a name has been read and is reset when a value is read
-  // No need to stack this, when we pop, we know we have to read a new name
-  private var pendingName: String? = null
 
-  init {
-    /**
-     * We only allow Maps at the root of the Json. Initialize the state accordingly
-     */
-    push(OrderedMap(listOf(Entry("root", root))))
-    pendingName = "root"
-  }
+  /**
+   * Can contain either:
+   * - an Int representing the next index to be read in a List
+   * - a String representing the current key to be read in a Map
+   * - null if peekedToken is BEGIN_OBJECT
+   */
+  private val path = arrayOfNulls<Any>(MAX_STACK_SIZE)
 
-  private fun push(data: Any) {
-    check(stackSize < MAX_STACK_SIZE) {
-      "Nesting too deep"
-    }
+  /**
+   * The current object memorized in case we need to rewind
+   */
+  private var containerStack = arrayOfNulls<Map<String, Any?>>(MAX_STACK_SIZE)
+  private val iteratorStack = arrayOfNulls<Iterator<*>>(MAX_STACK_SIZE)
+  private val nameIndexStack = IntArray(MAX_STACK_SIZE)
 
-    dataStack[stackSize] = data
-    indexStack[stackSize] = 0
-    nameIndexStack[stackSize] = 0
+  private var stackSize = 0
 
-    stackSize++
-    pendingName = null
+  init {
+    peekedToken = JsonReader.Token.BEGIN_OBJECT
+    peekedData = root
   }
 
-  private fun pop() {
-    // Allow garbage collection
-    dataStack[stackSize] = null
-    stackSize--
-    pendingName = null
+  private fun anyToToken(any: Any?) = when (any) {
+    null -> JsonReader.Token.NULL
+    is List<*> -> JsonReader.Token.BEGIN_ARRAY
+    is Map<*, *> -> JsonReader.Token.BEGIN_OBJECT
+    is Int -> JsonReader.Token.NUMBER
+    is Long -> JsonReader.Token.NUMBER
```

Comment:
Would it make sense to return `JsonReader.Token.LONG` for this one?

Reply:
Looks like it 👍 . I must admit I'm not 100% clear the implications of this as `nextInt()`, `nextDouble()`, etc.. do some coercion. But it's certainly better to hint at long there.